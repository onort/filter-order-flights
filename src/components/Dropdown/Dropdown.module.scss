@import "../../styles/main.scss";

.container {
  display: flex;
  align-items: center;
}

.title {
  display: block;
  margin-right: map-get($spacers, normal);
  color: $grey1;
  font-family: $font-family-heading;
  font-size: map-get($heading-sizes, h4);
  font-weight: 600;
}

.selected {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.arrow {
  width: 0;
  height: 0;
  border-top: px-to-rem(5px) solid $grey3;
  border-right: px-to-rem(5px) solid transparent;
  border-left: px-to-rem(5px) solid transparent;
}

.arrowUp {
  border-top: none;
  border-bottom: px-to-rem(5px) solid $grey3;
}

.dropdown {
  position: relative;
  display: block;
  width: px-to-rem(230px);
  padding: map-get($spacers, small) map-get($spacers, normal);
  margin-left: 1rem;
  background-color: $white;
  color: $grey3;
  cursor: pointer;
  font-family: $font-family-heading;
  font-size: map-get($heading-sizes, h4);
  font-weight: 600;
}

.dropdownList {
  position: absolute;
  z-index: 2;
  top: px-to-rem(30px);
  left: 0;
  display: block;
  width: px-to-rem(230px);
  padding: 0;
  background-color: $white;
  border-radius: px-to-rem(2px);
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}

.dropdownItem {
  padding: map-get($spacers, small) map-get($spacers, normal);
  cursor: pointer;
  list-style: none;

  &:hover {
    background-color: $grey5;
  }
}
