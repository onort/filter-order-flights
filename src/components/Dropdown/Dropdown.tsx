import React, { Component } from "react"
import cx from "classnames"

import styles from "./Dropdown.module.scss"
import { OrderingOption } from "../../types"

interface Props {
  onSelect: (value: string) => void
  orderingOtions: OrderingOption[]
  selectedOrder: OrderingOption
  title?: string
}

interface State {
  open: boolean
}

class Dropdown extends Component<Props, State> {
  public state = {
    open: false
  }

  public handleSelect = (value: string) => {
    this.props.onSelect(value)
    this.toggle()
  }

  public toggle = () => {
    this.setState(state => ({ open: !state.open }))
  }

  public render() {
    const { title, orderingOtions, selectedOrder } = this.props
    const { open } = this.state
    return (
      <div className={styles.container}>
        {title && <span className={styles.title}>{title}</span>}
        <div className={styles.dropdown}>
          <div className={styles.selected} onClick={this.toggle}>
            <div className={styles.selectedItem}>{selectedOrder.text}</div>
            <div className={cx(styles.arrow, { [styles.arrowUp]: open })} />
          </div>
          {open && (
            <ul className={styles.dropdownList}>
              {orderingOtions.map(option => (
                <li
                  key={option.value}
                  className={styles.dropdownItem}
                  onClick={this.handleSelect.bind(this, option.value)}
                >
                  {option.text}
                </li>
              ))}
            </ul>
          )}
        </div>
      </div>
    )
  }
}

export default Dropdown
