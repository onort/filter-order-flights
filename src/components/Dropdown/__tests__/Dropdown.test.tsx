import React from "react"
import { shallow } from "enzyme"

import Dropdown from "../"

const testTitle = "Dropdown"
const mockOnSelect = jest.fn
const orderingOptions = [
  { text: "Ucuzdan Pahalıya", value: "minPrice" },
  { text: "Pahalıdan Ucuza", value: "maxPrice" },
  { text: "Tarihe Göre Yakın", value: "closest" },
  { text: "Tarihe Göre Uzak", value: "farthest" }
]
const selectedOrder = orderingOptions[0]

describe("<Dropdown />", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(
      <Dropdown
        title={testTitle}
        onSelect={mockOnSelect}
        orderingOtions={orderingOptions}
        selectedOrder={selectedOrder}
      />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render title when passed", () => {
    const wrapper = shallow(
      <Dropdown
        title={testTitle}
        onSelect={mockOnSelect}
        orderingOtions={orderingOptions}
        selectedOrder={selectedOrder}
      />
    )
    expect(wrapper.find(".title").text()).toBe(testTitle)
  })

  it("should show select list when clicked", () => {
    const wrapper = shallow(
      <Dropdown
        title={testTitle}
        onSelect={mockOnSelect}
        orderingOtions={orderingOptions}
        selectedOrder={selectedOrder}
      />
    )
    expect(wrapper.find(".dropdownList").exists()).toBe(false)
    wrapper.find(".selected").simulate("click")
    expect(wrapper.find(".dropdownList").exists()).toBe(true)
  })

  it("should have correct amount of list items", () => {
    const wrapper = shallow(
      <Dropdown
        title={testTitle}
        onSelect={mockOnSelect}
        orderingOtions={orderingOptions}
        selectedOrder={selectedOrder}
      />
    )
    wrapper.find(".selected").simulate("click")
    expect(wrapper.find("li")).toHaveLength(orderingOptions.length)
  })

  it("should fire onSelect when a list item is clicked", () => {
    const onSelect = jest.fn()
    const wrapper = shallow(
      <Dropdown
        title={testTitle}
        onSelect={onSelect}
        orderingOtions={orderingOptions}
        selectedOrder={selectedOrder}
      />
    )
    wrapper.find(".selected").simulate("click")
    wrapper
      .find("li")
      .at(2)
      .simulate("click")
    expect(onSelect).toBeCalledTimes(1)
    expect(onSelect).toBeCalledWith(orderingOptions[2].value)
  })
})
