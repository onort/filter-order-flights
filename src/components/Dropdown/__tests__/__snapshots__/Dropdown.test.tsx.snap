// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    onSelect={[Function]}
    orderingOtions={
      Array [
        Object {
          "text": "Ucuzdan Pahalıya",
          "value": "minPrice",
        },
        Object {
          "text": "Pahalıdan Ucuza",
          "value": "maxPrice",
        },
        Object {
          "text": "Tarihe Göre Yakın",
          "value": "closest",
        },
        Object {
          "text": "Tarihe Göre Uzak",
          "value": "farthest",
        },
      ]
    }
    selectedOrder={
      Object {
        "text": "Ucuzdan Pahalıya",
        "value": "minPrice",
      }
    }
    title="Dropdown"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="title"
        >
          Dropdown
        </span>,
        <div
          className="dropdown"
        >
          <div
            className="selected"
            onClick={[Function]}
          >
            <div
              className="selectedItem"
            >
              Ucuzdan Pahalıya
            </div>
            <div
              className="arrow"
            />
          </div>
        </div>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Dropdown",
          "className": "title",
        },
        "ref": null,
        "rendered": "Dropdown",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="selected"
              onClick={[Function]}
            >
              <div
                className="selectedItem"
              >
                Ucuzdan Pahalıya
              </div>
              <div
                className="arrow"
              />
            </div>,
            false,
          ],
          "className": "dropdown",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="selectedItem"
                >
                  Ucuzdan Pahalıya
                </div>,
                <div
                  className="arrow"
                />,
              ],
              "className": "selected",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Ucuzdan Pahalıya",
                  "className": "selectedItem",
                },
                "ref": null,
                "rendered": "Ucuzdan Pahalıya",
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "arrow",
                },
                "ref": null,
                "rendered": null,
                "type": "div",
              },
            ],
            "type": "div",
          },
          false,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="title"
          >
            Dropdown
          </span>,
          <div
            className="dropdown"
          >
            <div
              className="selected"
              onClick={[Function]}
            >
              <div
                className="selectedItem"
              >
                Ucuzdan Pahalıya
              </div>
              <div
                className="arrow"
              />
            </div>
          </div>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Dropdown",
            "className": "title",
          },
          "ref": null,
          "rendered": "Dropdown",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="selected"
                onClick={[Function]}
              >
                <div
                  className="selectedItem"
                >
                  Ucuzdan Pahalıya
                </div>
                <div
                  className="arrow"
                />
              </div>,
              false,
            ],
            "className": "dropdown",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="selectedItem"
                  >
                    Ucuzdan Pahalıya
                  </div>,
                  <div
                    className="arrow"
                  />,
                ],
                "className": "selected",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Ucuzdan Pahalıya",
                    "className": "selectedItem",
                  },
                  "ref": null,
                  "rendered": "Ucuzdan Pahalıya",
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "arrow",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "div",
                },
              ],
              "type": "div",
            },
            false,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`<Dropdown /> should match snapshot 2`] = `"<div class=\\"container\\"><span class=\\"title\\">Dropdown</span><div class=\\"dropdown\\"><div class=\\"selected\\"><div class=\\"selectedItem\\">Ucuzdan Pahalıya</div><div class=\\"arrow\\"></div></div></div></div>"`;
